File: virtual-memory.py
from bitstring import BitArray
import threading
import time

#64mb ram
#page 4mb

MB = 1024 * 1024

memory = 64 * MB

buffer = bytearray(memory)

isinmemory = BitArray(256)
struct = [] # 256 * 3 byte
#TODO: add own: just arry no list

#  64MB 4MB 4096 4194304
#  read 2000s byte => page 0
#  read 5000000s byte => page 1

# Used RAM
# is in memory: 32 byte (256bit)
# 256 *
# lastused => 2byte => 18h
# where start address: 1 byte 0 - 15 255=>unmapped
# 800 byte

currenttimer = 0

def addressToPageNr(address):
    for i in range(256):
        min = i * 4 * MB
        max = (i+1) * 4 * MB
        if address >= min and address < max:
            return i

def getAddressPage(pagenr):
    startaddr = getStartAddressInMemory(pagenr)
    min = startaddr * 4 * MB
    max = min + 4 * MB
    return buffer[min:max]

def isPageInMemory(pagenr):
    return isinmemory[pagenr]

def setIsPageInMemory(pagenr, isin):
    isinmemory[pagenr] = int(isin)

def isAddressInMemory(address):
    return isPageInMemory(addressToPageNr(address))

def getLastUsed(pagenr):
    byte1 = struct[pagenr][0].to_bytes(1, byteorder="big")
    byte2 = struct[pagenr][1].to_bytes(1, byteorder="big")
    return int.from_bytes(byte1 + byte2, "big")

def setLastUsed(pagenr, time):
    byte12 = time.to_bytes(2, byteorder="big")
    byte3 = struct[pagenr][2].to_bytes(1, byteorder="big")
    byte123 = byte12[0].to_bytes(1, byteorder="big") + byte12[1].to_bytes(1, byteorder="big") + byte3[0].to_bytes(1, byteorder="big")
    struct[pagenr] = byte123

def getStartAddressInMemory(pagenr):
    byte3 = struct[pagenr][2].to_bytes(1, byteorder="big")
    return int.from_bytes(byte3, "big")

def setStartAddressInMemory(pagenr, address):
    byte1 = struct[pagenr][0].to_bytes(1, byteorder="big")
    byte2 = struct[pagenr][1].to_bytes(1, byteorder="big")
    byte3 = address.to_bytes(1, byteorder="big")
    byte123 = byte1[0].to_bytes(1, byteorder="big") + byte2[0].to_bytes(1, byteorder="big") + byte3[0].to_bytes(1, byteorder="big")
    struct[pagenr] = byte123

def isAddressPlusLengthInSamePage(address, length):
    page = addressToPageNr(address)
    pagemaxaddress = (page * 4 * MB) + (4 * MB - 1)
    return address + length <= pagemaxaddress

def isPageEmpty(pagenr):
    for byte in getAddressPage(pagenr):
        if byte != 0:
            return False
    return True
    

def handlePageFault(pagenr):
    unusedpage = getUnusedPage()
    realunused = getStartAddressInMemory(unusedpage)
    if not isPageEmpty(unusedpage):
        writePageToDisk(unusedpage)
    else:
        setIsPageInMemory(unusedpage, False)
        setStartAddressInMemory(unusedpage, 255)
    loadPageFromDisk(pagenr, realunused)

def readFromMemory(address):
    page = addressToPageNr(address)
    if isPageInMemory(page):
        newaddr = address - (4 * MB * page)
        setLastUsed(page, currenttimer)
        return getAddressPage(page)[newaddr]
    else:
        handlePageFault(page)
        return readFromMemory(address)

def writeToMemory(address, byte):
    page = addressToPageNr(address)
    if isPageInMemory(page):
        diffaddr = address - (4 * MB * page)
        newaddr = getStartAddressInMemory(page) * 4 * MB + diffaddr
        setLastUsed(page, currenttimer)
        buffer[newaddr] = byte
    else:
        handlePageFault(page)
        writeToMemory(address, byte)


def writeFileToDisk(name, array):
    nfile = open(name, 'wb')
    nfile.write(array)
    nfile.close()

def readFileFromDisk(name):
    nfile = open(name, 'rb')   
    buff = nfile.read()
    nfile.close()
    return buff

def writeBytesToBuffer(array, address):
    for i in range(len(array)):
        buffer[address + i] = array[i]

def writePageToDisk(pagenr):
    setIsPageInMemory(pagenr, False)
    writeFileToDisk(str(pagenr), getAddressPage(pagenr))
    setStartAddressInMemory(pagenr, 255)

def loadPageFromDisk(pagenr, startaddr):
    try:
        buf = readFileFromDisk(str(pagenr))
    except FileNotFoundError:
        buf = bytearray(4 * MB)
    writeBytesToBuffer(buf, startaddr * 4 * MB)
    setIsPageInMemory(pagenr, True)
    setStartAddressInMemory(pagenr, startaddr)

def getUnusedPage():
    longesttime = 0
    longest = -1
    for i in range(256):
        if isPageInMemory(i):
            timediff = currenttimer - getLastUsed(i)
            if timediff > longesttime:
                longesttime = timediff
                longest = i

    return longest

def incTimer():
    global currenttimer
    global thread
    currenttimer += 1
    thread = threading.Timer(0.1, incTimer) # old 1.0
    thread.start()

def setup():
    for i in range(16): #0-15
        isinmemory[i] = 1

    for i in range(1, 256): # 256 * 3 byte
        struct.append(b'\x00\x00\xff')

    for i in range(16):
        setStartAddressInMemory(i, i)
        setLastUsed(i, currenttimer)

    incTimer()
    

setup()

def showmappings():
    buf = ""
    for i in range(25):
        buf += str(i) + ":" + str(getStartAddressInMemory(i)) + " " 
    print(buf)

import random


def getActiveMappings():
    count = 0
    for i in range(256):
        if isPageInMemory(i):
            count += 1
    print(count)


def eversecond():
    global thread1s
    global time1s

    times = 0.0
    time = 0.0
    stby = 0.0

    for i in range(0, random.randint(4,12)):
        stby = currenttimer
        writeToMemory(i * 4 * MB + 1024, random.randint(0,255))
        time = currenttimer - stby
        if time > 0:
            print("time1s: " + str(time))
    
    getActiveMappings()

    thread1s = threading.Timer(1.0, eversecond)
    thread1s.start()


def every30second():
    global thread30s   
    global time30s

    times = 0.0
    time = 0.0
    stby = 0.0

    for i in range(random.randint(7,14),26):
        stby = currenttimer
        writeToMemory(i * 4 * MB + 1024, random.randint(0,255))
        time = currenttimer - stby
        print("time30s: " + str(time))
    

    thread30s = threading.Timer(30.0, every30second)
    thread30s.start()

def everyminute():

    global thread60s    
    global time60s

    times = 0.0
    time = 0.0
    stby = 0.0

    for i in range(16, random.randint(18,78)):
        stby = currenttimer
        writeToMemory(i * 4 * MB + 1024, random.randint(0,255))
        time = currenttimer - stby
        print("time60s: " + str(time))

    buf = ""
    for i in range(25):
        buf += str(i) + ":" + str(getStartAddressInMemory(i)) + " " 
    print(buf)

    thread60s = threading.Timer(60.0, everyminute)
    thread60s.start()


#eversecond()
#every30second()
#everyminute()

input("done?")


getActiveMappings()
writeToMemory(1024, 255)
getActiveMappings()
writeToMemory(len(buffer) + 1024, 255)
getActiveMappings()
for i in range(16, random.randint(18,78)):
    writeToMemory(i * 4 * MB + 1024, random.randint(0,255))
getActiveMappings()



thread.cancel()

File: virtual-memory.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <pthread.h>

#include "virtual-memory.h"

//Virtual Memory
//RAM Size: 64MB
//Block Size: 4MB
//Used Memory:
//  isinmemory:     256 * 1bit => 32 bytes
//  lastused:       256 * 4bytes
//  mappedaddress:  256 * 1byte
//  padding:        256 * 3byte
//  Total:          2080bytes

#define MB  1024 * 1024
#define RAM_SIZE    64 * MB


volatile uint32_t timer = 0;

char *buffer;

char *isinmemory;
struct page *pages;

struct page
{
    uint32_t lastused;
    uint8_t mappedaddress;
};



int main(){

    pthread_t thread1, thread2;

    printf("starting threads...\n");
    pthread_create(&thread1, NULL, incTimer, NULL);
    pthread_create(&thread2, NULL, setup, NULL);

    pthread_join(thread2, NULL);
    printf("done\n");

    return 0;

}



void *setup(void *vargp){

    buffer = calloc(RAM_SIZE, 1);

    isinmemory = (char *) calloc(32, 1);
    pages = (struct page *) calloc(256, sizeof(struct page));

    for(int i=0; i<256; i++){
        if(i<16){
            setBit(isinmemory, i);
            pages[i].mappedaddress = i;
        }else{
            pages[i].mappedaddress = 0xff;
        }
    }

    printf("start test program\n");

    writeToMemory(1024, 0x88);
    setLastUsed(0, 0);
    writeToMemory(18 * 4 * MB + 1024, 0xfe);


    printf("%x\n", readFromMemory(1024));
    printf("%x\n", readFromMemory(18 * 4 * MB + 1024));

    return NULL;
}



//Main Functions

void handlePageFault(uint8_t pagenr){
    uint8_t unusedpage = getUnusedPage();
    uint8_t realunused = getMappedAddressInMemory(unusedpage);
    if(!isPageEmpty(unusedpage)){
        writePageToDisk(unusedpage);
    }else{
        setIsPageInMemory(unusedpage, 0);
        setMappedAddressInMemory(unusedpage, 255);
    }
    loadPageFromDisk(pagenr, realunused);
}

char readFromMemory(uint32_t address){
    uint8_t page = addressToPageNr(address);
    if(isPageInMemory(page)){
        uint32_t offset = address - (4 * MB * page);
        setLastUsed(page, timer);
        char *buffer = getAddressPage(page);
        char byte = buffer[offset];
        free(buffer);
        return byte;
    }else{
        handlePageFault(page);
        return readFromMemory(address);
    }
}

void writeToMemory(uint32_t address, char byte){
    uint8_t page = addressToPageNr(address);
    if(isPageInMemory(page)){
        uint32_t offset = address - (4 * MB * page);
        uint32_t newaddr = getMappedAddressInMemory(page) * 4 * MB + offset;
        setLastUsed(page, timer);
        buffer[newaddr] = byte;
    }else{
        handlePageFault(page);
        writeToMemory(address, byte);
    }
}



//Page swapping function

void writePageToDisk(uint8_t pagenr){
    char str_pagenr[3];
    sprintf(str_pagenr, "%d", pagenr);
    setIsPageInMemory(pagenr, 0);
    char *buffer = getAddressPage(pagenr);
    writeFileToDisk(str_pagenr, 4 * MB, buffer);
    setMappedAddressInMemory(pagenr, 255);
    free(buffer);
}

void loadPageFromDisk(uint8_t pagenr, uint8_t startaddr){
    char str_pagenr[3];
    sprintf(str_pagenr, "%d", pagenr);
    char *buf = readFileFromDisk(str_pagenr, 4 * MB);
    writeBytesToBuffer(buf, 4 * MB, startaddr * 4 * MB);
    setIsPageInMemory(pagenr, 1);
    setMappedAddressInMemory(pagenr, startaddr);
    free(buf);
}



//File functions

void writeFileToDisk(char *name, uint32_t length, char *array){
    FILE *nfile = fopen(name, "wb");
    fwrite(array, length, 1, nfile);
    fclose(nfile);
}

//Memory needs to be free'd
char *readFileFromDisk(char *name, uint32_t length){
    char *array;
    FILE *nfile = fopen(name, "rb");
    if(nfile == NULL){
        array = (char *) calloc(length, 1);
    }else{
        array = (char *) malloc(length);
        fread(array, length, 1, nfile);
        fclose(nfile);
    }
    return array;
}



//Page functions

uint8_t addressToPageNr(uint32_t address){
    for(int i=0;i<256;i++){
        uint32_t min = i * 4 * MB;
        uint32_t max = (i+1) * 4 * MB;
        if(address >= min && address < max)
            return i;
    }
    printf("Address out of range\n");
    return 0;
}

//Memory needs to be free'd
char *getAddressPage(uint8_t pagenr){
    uint8_t startaddr = getMappedAddressInMemory(pagenr);
    uint32_t min = startaddr * 4 * MB;
    uint32_t max = min + 4 * MB;
    return sliceArray(buffer, 4 * MB, min, max);
}

uint8_t isPageEmpty(uint8_t pagenr){
    char *page = getAddressPage(pagenr);
    uint8_t isempty = 1;
    for(int i=0; i<(4 * MB); i++){
        if(page[i] != 0){
            isempty = 0;
            break;
        }
    }
    free(page);
    return isempty;
}

uint8_t getUnusedPage(){
    uint32_t longesttime = 0;
    uint8_t longest = 0;
    for(int i=0; i<256; i++){
        if(isPageInMemory(i)){
            uint32_t timediff = timer - getLastUsed(i);
            if(timediff > longesttime){
                longesttime = timediff;
                longest = i;
            }
        }
    }
    if(longest == 0 && !isPageInMemory(0))
        printf("This should not have happened\n"); //TODO
    return longest;
}



//Basic utility functions

uint8_t isPageInMemory(uint8_t pagenr){
    return getBit(isinmemory, pagenr);
}

void setIsPageInMemory(uint8_t pagenr, uint8_t isin){
    if(isin)
        setBit(isinmemory, pagenr);
    else
        clearBit(isinmemory, pagenr);
}

uint8_t isAddressInMemory(uint32_t address){
    return isPageInMemory(addressToPageNr(address));
}

uint32_t getLastUsed(uint8_t pagenr){
    return pages[pagenr].lastused; 
}

void setLastUsed(uint8_t pagenr, uint32_t time){
    pages[pagenr].lastused = time;
}

uint8_t getMappedAddressInMemory(uint8_t pagenr){
    return pages[pagenr].mappedaddress;
}

void setMappedAddressInMemory(uint8_t pagenr, uint8_t address){
    pages[pagenr].mappedaddress = address;
}

void writeBytesToBuffer(char *array, uint32_t length, uint32_t address){
    for(int i=0; i<length; i++){
        buffer[address + i] = array[i];
    }
}



//Timer

void *incTimer(void *vargp){
    while(1){
        usleep(1000);
        timer++;
    }
    return NULL;
}



//Basic help functions

//Memory needs to be free'd
char *sliceArray(char *array, uint32_t arraysize, uint32_t pos1, uint32_t pos2){ 
    char *returnarray = (char *) malloc((pos2 - pos1) * sizeof(char));
    for(int i=pos1; i<pos2; i++)
        returnarray[i - pos1] = array[i];
    return returnarray;
}

int getIndexFromArray(char *array, int nr){
    for(int i=0; i<32; i++)
        if(nr < ((i + 1) * 8))
            return i;
    printf("Number out of range\n");
    return 0;
}

void setBit(char *array, int nr){
    int index = getIndexFromArray(array, nr);
    nr -= (index * 8);
    array[index] |= (1 << nr);
}

void clearBit(char *array, int nr){
    int index = getIndexFromArray(array, nr);
    nr -= (index * 8);
    array[index] &= ~(1 << nr);
}

int getBit(char *array, int nr){
    int index = getIndexFromArray(array, nr);
    nr -= (index * 8);
    return (array[index] & (1 << nr)) == (1 << nr);
}
